osquery/filesystem/filesystem.cpp

```diff
@@ -57,23 +61,45 @@ Status writeTextFile(const fs::path& path,
 }
 
 Status readFile(const fs::path& path, std::string& content) {
-  auto path_exists = pathExists(path);
-  if (!path_exists.ok()) {
-    return path_exists;
+  struct stat file;
+  if (lstat(path.string().c_str(), &file) == 0) {
+    if (file.st_uid != 0 && !FLAGS_read_user_links) {
+      return Status(1, "User link reads disabled");
+    }
+  } else if (stat(path.string().c_str(), &file) < 0) {
+    return Status(1, "Cannot access path: " + path.string());
   }
 
-  std::stringstream buffer;
-  fs::ifstream file_h(path);
-  if (file_h.is_open()) {
-    buffer << file_h.rdbuf();
-    if (file_h.bad()) {
-      return Status(1, "Error reading file: " + path.string());
+  // Apply the max byte-read based on file/link target ownership.
+  size_t read_max = (file.st_uid == 0)
+                        ? FLAGS_read_max
+                        : std::min(FLAGS_read_max, FLAGS_read_user_max);
+  std::ifstream is(path.string(), std::ifstream::binary);
+  if (!is) {
+    return Status(1, "Error reading file: " + path.string());
+  }
+
+  // Attempt to read the file size.
+  ssize_t size = file.st_size;
+
+  // Erase/clear provided string buffer.
+  content.erase();
+  if (file.st_size > read_max) {
+    return Status(1, "File exceeds read limits");
+  }
+
+  if (size == -1 || size == 0) {
```

Comment:
I'm not sure if the stream and move + assign is any faster/slower than an allocate and read. It might be helpful to remove the option for allocate and read if it has negligible or worse performance.